### PR TITLE
[Translation][Chinese] post-09 (edition-2)

### PR DESCRIPTION
**Pull request by weijiew**

### CODE REVIEW

**Comment by phil-opp** [comment]
Thanks a lot for submitting! Is this ready for a review?

**Comment by weijiew** [comment]
How do I solve this problem? The error message is "operation timed out", resubmit? I am new to the process and not familiar with it.

File: blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md [hunk 1]
```diff
@@ -0,0 +1,1012 @@
...
+
+这篇文章最后说，我们[不能从我们的内核中访问页表][end of previous post]，因为它们存储在物理内存中，而我们的内核已经在虚拟地址上运行。
+
+这篇文章探讨了使页表框能够被内核访问的不同方法。我们将讨论每种方法的优点和缺点，然后为我们的内核决定一种方法。
+
+[end of previous post]: @/edition-2/posts/08-paging-introduction/index.md#accessing-the-page-tables
+
+为了实现这个方法，我们需要 bootloader 的支持，所以首先要配置它。之后将实现一个遍历页表层次结构的函数，以便将虚拟地址转换为物理地址。最后，我们学习如何在页表中创建新的映射，以及如何为创建新的页表找到未使用的内存框。
+
+## Accessing Page Tables
```
Comment (phil-opp) [inline blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md:45]: You can keep these headings in Chinese. For the anchor links you can use the pinyin romanization. For example, the generated anchor for `链接器错误` is `#lian-jie-qi-cuo-wu`. I'm happy to help you with that if you have problems.

Comment (weijiew) [reply blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md:45]: Thank you very much, I will try to fix this now.

**Comment by weijiew** [comment]
All checks have passed, can this pr be merged?

File: blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md (hunk 1 above)
Comment (phil-opp) [reply blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md]: Thanks!

**Comment by phil-opp** [comment]
I don't speak Chinese myself, so I have to look for someone who can review this.

cc @liuyuran @JiangengDong @Byacrya Maybe one of you could do a quick review of this PR? Thanks!

File: blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md [hunk 2]
```diff
@@ -0,0 +1,1010 @@
...
+[at the bottom]: #comments
+<!-- fix for zola anchor checker (target is in template): <a id="comments"> -->
+[post branch]: https://github.com/phil-opp/blog_os/tree/post-09
+
+<!-- toc -->
+
+
+## 介绍
+
+上一篇[文章]已经对分页的概念做了介绍。通过比较分页和分段来证明分页的优势，然后解释了分页和页表如何工作，最后介绍了`x86_64`的4级页表设计。此时 bootloader 已经为内核建立了一个页表层次结构，这意味着内核已经在虚拟地址上运行。这样做提高了安全性，因为非法的内存访问会导致页面故障异常，而不是修改任意的物理内存。
```
Comment (liuyuran) [inline blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md:33]: well, I'm not good at review translations, but the `[文章]` seems like a dead link?
suggestion: `[前文]已经对分页的概念做了介绍。`

File: blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md [hunk 3]
```diff
@@ -0,0 +1,1010 @@
...
+由于每个物理地址都可以通过添加`physical_memory_offset`来访问，`physical_memory_offset`地址的翻译本身应该指向物理地址`0`。然而，翻译失败了，因为映射使用了巨大的页面来提高效率，而我们的实现还不支持。
+
+### 使用 `OffsetPageTable`
+
+将虚拟地址转换为物理地址是操作系统内核中的一项常见任务，因此`x86_64`内核为它提供了一个抽象。这个实现已经支持巨大的页面和除了 "translate_addr "之外的其他几个页表函数，所以我们将在下面使用它，而不是在我们自己的实现中添加巨大的页面支持。
+
+抽象的基础是两个特征，它们定义了各种页表映射功能。
+
+- [`Mapper`] 特质在页面大小上是通用的，并提供对页面进行操作的函数。例如[`translate_page`]，它将一个给定的页面翻译成相同大小的框架，以及[`map_to`]，它在页面表中创建一个新的映射。
+- `Translate`]特性提供了与多个页面大小有关的函数，如[`translate_addr`]或一般[`translate`]。
```
Comment (liuyuran) [inline blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md:615]: lost a `[` at front of line

File: blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md [hunk 4]
```diff
@@ -0,0 +1,1010 @@
...
+
+[`Mapper`]: https://docs.rs/x86_64/0.14.2/x86_64/structures/paging/mapper/trait.Mapper.html
+[`translate_page`]: https://docs.rs/x86_64/0.14.2/x86_64/structures/paging/mapper/trait.Mapper.html#tymethod.translate_page
+[`map_to`]: https://docs.rs/x86_64/0.14.2/x86_64/structures/paging/mapper/trait.Mapper.html#method.map_to
+[`Translate`]: https://docs.rs/x86_64/0.14.2/x86_64/structures/paging/mapper/trait.Translate.html
+[`translate_addr`]: https://docs.rs/x86_64/0.14.2/x86_64/structures/paging/mapper/trait.Translate.html#method.translate_addr
+[`translate`]: https://docs.rs/x86_64/0.14.2/x86_64/structures/paging/mapper/trait.Translate.html#tymethod.translate
+
+特质只定义接口，不提供任何实现。`x86_64`板块目前提供了三种类型来实现不同要求的特征。[`OffsetPageTable`] 类型假设完整的物理内存被映射到虚拟地址空间的某个偏移处。[`MappedPageTable']更灵活一些。它只要求每个页表帧在一个可计算的地址处被映射到虚拟地址空间。最后，[递归页表]类型可以用来通过[递归页表](#di-gui-ye-biao)访问页表框架。
```
Comment (liuyuran) [inline blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md:624]: may be the [\`MappedPageTable'] use a wrong symbol ?
should be [\`MappedPageTable\`]

File: blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md [hunk 5]
```diff
@@ -0,0 +1,1010 @@
...
+        // 从起始地址创建 `PhysFrame`  类型 
+        frame_addresses.map(|addr| PhysFrame::containing_address(PhysAddr::new(addr)))
+    }
+}
+```
+
+这个函数使用迭代器组合方法将初始的`MemoryMap'转化为可用的物理帧的迭代器。
+
+- 首先，我们调用`iter`方法，将内存映射转换为[`MemoryRegion`]s的迭代器。
+- 然后我们使用[`filter`]方法跳过任何保留或其他不可用的区域。Bootloader为它创建的所有映射更新了内存地图，所以被我们的内核使用的帧（代码、数据或堆栈）或存储启动信息的帧已经被标记为`InUse'或类似的。因此，我们可以确定 "可使用 "的帧没有在其他地方使用。
```
Comment (liuyuran) [inline blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md:927]: \`InUse' should be \`InUse\`
and just a suggestion, not a mistake,  `确定 "可使用 "的帧` -> `确定 "可使用" 的帧`

File: blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md [hunk 6]
```diff
@@ -0,0 +1,1010 @@
...
+            .map(|r| r.range.start_addr()..r.range.end_addr());
+        // 转化为一个帧起始地址的迭代器
+        let frame_addresses = addr_ranges.flat_map(|r| r.step_by(4096));
+        // 从起始地址创建 `PhysFrame`  类型 
+        frame_addresses.map(|addr| PhysFrame::containing_address(PhysAddr::new(addr)))
+    }
+}
+```
+
+这个函数使用迭代器组合方法将初始的`MemoryMap'转化为可用的物理帧的迭代器。
```
Comment (liuyuran) [inline blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md:924]: \`MemoryMap' should be \`MemoryMap\`

File: blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md [hunk 7]
```diff
@@ -0,0 +1,1010 @@
...
+
+此时，我们应该再次删除`create_example_mapping`函数，以避免意外地调用未定义的行为，正如[上面](#create_example_mapping 函数)所解释的那样。
+
+## 总结
+
+在这篇文章中，我们了解了访问页表物理框架的不同技术，包括 identity mapping 、完整物理内存的映射、临时映射和递归页表。我们选择了映射完整的物理内存，因为它简单、可移植，而且功能强大。
+
+我们不能在没有页表访问的情况下从我们的内核映射物理内存，所以我们需要bootloader的支持。`bootloader`板块支持通过可选的 cargo 板块功能创建所需的映射。它以"&BootInfo "参数的形式将所需信息传递给我们的内核。
+
+对于我们的实现，我们首先手动遍历页表以实现翻译功能，然后使用`x86_64'板块的`MappedPageTable`类型。我们还学习了如何在页表中创建新的映射，以及如何在引导程序传递的内存映射之上创建必要的 "FrameAllocator"。
```
Comment (liuyuran) [inline blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md:1003]: \`x86_64' should be \`x86_64\`

**Review by liuyuran — COMMENTED** [review]
and only a suggestion, I think `Identity Mapping` means `直接映射` but not sure ...

File: blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md (hunk 7 above)
Comment (weijiew) [reply blog/content/edition-2/posts/09-paging-implementation/index.zh-CN.md]: Based on your suggestion, I have corrected the article and thank you for reviewing it.

**Comment by phil-opp** [comment]
@liuyuran Thanks a lot for reviewing! Is this PR ready to be merged from your side?

**Comment by liuyuran** [comment]
> @liuyuran Thanks a lot for reviewing! Is this PR ready to be merged from your side?

no problem I think? @weijiew the second question seems that up to you.

**Comment by phil-opp** [comment]
Ok great, then let's get this merged! Thanks again @weijiew!

**Pull request merged**